File: 49.group-anagrams.py
#
# @lc app=leetcode id=49 lang=python3
#
# [49] Group Anagrams
#

# Time:  O(n * aloga), a is max anagram string length
# Space: O(n)

from typing import List
from collections import defaultdict


# @lc code=start
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams_map = defaultdict(list)
        for str in strs:
            anagrams_map["".join(sorted(str))].append(str)
        return list(anagrams_map.values())


# @lc code=end

File: 49.group-anagrams.js
/*
 * @lc app=leetcode id=49 lang=javascript
 *
 * [49] Group Anagrams
 */

// @lc code=start
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  const lookup = {};
  for (const s of strs) {
    const sorted = s.split("").sort().join("");
    if (!lookup[sorted]) {
      lookup[sorted] = [];
    }
    lookup[sorted].push(s);
  }
  return Object.values(lookup);
};
// @lc code=end
